Added forgot password recovery

TarasOrlovskyi opened the pull request:
1. Added V1_7__recovery_password.sql for create recovery_password table
   and change link_to_main_page for Triplevision in db
2. Added shops small logos in img folder
3. Into DefaultUserPostService added 2 methods: AddRecoveryUserToken and
   GetRecoveryUserId
4. In UserService and UserDao added findById method
5. In UserPostDao added updateRecoveryUserToken, addRecoveryUserToken, getRecoveryUserId and getRecoveryUserToken
6. Canged some css and html files in order to show warning message correctly and checking password/confirm password and validate email input
7. Added NewPasswordServlet in order to handle the request after the
   user click the link in the email
8. Into PageGenerator added two new paramether in Context
9. Added RecoveryPasswordServlet
10. In SecurityFilter added /newPassword as allowed link
11. Added Session in SignInServlet and SignUpServlet into doPost methods
12. Added into UserPostServicenext methods - addRecoveryUserToken and getRecoveryUserId

TarasOrlovskyi committed:
Added forgot password recovery

1. Added V1_7__recovery_password.sql for create recovery_password table
   and change link_to_main_page for Triplevision in db
2. Added shops small logos in img folder
3. Into DefaultUserPostService added 2 methods: AddRecoveryUserToken and
   GetRecoveryUserId
4. In UserService and UserDao added findById method
5. In UserPostDao added updateRecoveryUserToken, addRecoveryUserToken, getRecoveryUserId and getRecoveryUserToken
6. Canged some css and html files in order to show warning message correctly and checking password/confirm password and validate email input
7. Added NewPasswordServlet in order to handle the request after the
   user click the link in the email
8. Into PageGenerator added two new paramether in Context
9. Added RecoveryPasswordServlet
10. In SecurityFilter added /newPassword as allowed link
11. Added Session in SignInServlet and SignUpServlet into doPost methods
12. Added into UserPostServicenext methods - addRecoveryUserToken and getRecoveryUserId

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java:
```diff
@@ -90,4 +90,17 @@ public boolean update(String oldEmail, String newEmail, String newPassword, Stri
+    @Override
+    public Optional<User> findById(long id){
+        Optional<User> optionalUser = Optional.empty();
+        if (id > 0){
```
Ctrl+Alt+L

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/web/servlets/RecoveryPasswordServlet.java:
```diff
@@ -0,0 +1,103 @@
+        }
+        if (email != null && !email.isEmpty()){
+            Optional<User> user = userService.getByEmail(email);
+            if (user.isPresent()){
```
Ctrl+Alt+L

Yuliia-Sh replied:
DRY

Yuliia-Sh replied:
one servlet - one method in service

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java:
```diff
@@ -55,7 +61,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
                 User user = optionalUser.get();
-                HttpSession session = request.getSession(true);
+                session = request.getSession(true);
                 session.setMaxInactiveInterval(60 * 60 * 5);
```
Please remove magic number. Add it to application.properties and this class fields.

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/Starter.java:
```diff
@@ -111,6 +114,8 @@ public void run() {
         AboutServlet aboutServlet = new AboutServlet();
+        RecoveryPasswordServlet recoveryPasswordServlet = new RecoveryPasswordServlet(userService, userPostService, mailSender);
+        NewPasswordServlet newPasswordServlet = new NewPasswordServlet(userService, userPostService);
 
```
I think name ChangePasswordServlet could be better

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDao.java:
```diff
@@ -150,4 +153,24 @@ public boolean update(String email, User user) {
+            throw new RuntimeException(e);
+        }
+        return Optional.ofNullable(user);
+    }
```
It could be returned inside try block, and user declared also inside

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDao.java:
```diff
@@ -150,4 +153,24 @@ public boolean update(String email, User user) {
+    @Override
+    public Optional<User> findById(long id) {
+        User user = null;
+        try (Connection connection = dataSource.getConnection();
```
it looks ugly here

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserPostDao.java:
```diff
@@ -1,22 +1,30 @@
+            " WHERE user_id = ?";
+    private static final String FIND_USER_ID = "SELECT user_id FROM public.recovery_password" +
+            " WHERE token = ?";
 
```
I would not recommend use schema name in the SQL as tables can't be moved to another schema in the database without changing code

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserPostDao.java:
```diff
@@ -54,4 +62,103 @@ public boolean add(UserPost post) {
+        } catch (SQLException e) {
+            log.error("Error while updating recovery token to public.recovery_password", e);
+            throw new RuntimeException(e);
+        }
```
1) What is difference with these two exceptions - why different behavior? 
2) "public." looks ugly from my point of view

TarasOrlovskyi replied:
Perhaps PSQLExeption will be thrown if a user with such a login exists

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserPostDao.java:
```diff
@@ -54,4 +62,103 @@ public boolean add(UserPost post) {
+            if (result > 0) {
+                isUpdate = true;
+            }
+        } catch (PSQLException e) {
```
return (result > 0)
or even without that - it is updated if exception isn't thrown
why this complexity is needed?

TarasOrlovskyi replied:
Creating a table is neither an INSERT nor an UPDATE, so it's normal to receive 0 as no row(s) were affected. So it will be better additionally checking

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserPostService.java:
```diff
@@ -30,6 +32,26 @@ public boolean add(UserPost post) {
+            isAdded = userPostDao.addRecoveryUserToken(userId, recoveryToken);
+        } else {
+            isAdded = userPostDao.updateRecoveryUserToken(userId, recoveryToken);
+        }
```
I think insert or update can be done using one sql and one call to dao - could you please look at INSERT ON CONFLICT

TarasOrlovskyi replied:
You are right! Thanks, it is cool decision! :)

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/RecoveryPasswordServlet.java:
```diff
@@ -0,0 +1,103 @@
+public class RecoveryPasswordServlet extends HttpServlet {
+
+    private static final String RECOVERY_MESSAGE = "Hello, to change your password, follow this link: \nhttp://localhost:8080/newPassword?token=";
+
```
http://localhost:8080 shouldn't be here

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/NewPasswordServlet.java:
```diff
@@ -0,0 +1,105 @@
+                attributes.put("userRole", user.getRole().toString());
+            }
+        }
+        String token = request.getParameter("token");
```
perhaps this code 
HttpSession session = request.getSession(false);
        if (session != null) {
            User user = (User) session.getAttribute("user");
            if (user != null) {
                attributes.put("userRole", user.getRole().toString());
            }
        }
can be separated in the static helpful method that can be used some servlets, I am not sure here

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/NewPasswordServlet.java:
```diff
@@ -0,0 +1,105 @@
+        }
+        String token = request.getParameter("token");
+        long userId = userPostService.getRecoveryUserId(token);
+        if (userId > 0) {
```
getUserIdByRecoveryToken - I think this name is more readable

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/NewPasswordServlet.java:
```diff
@@ -0,0 +1,105 @@
+                attributes.put("userRole", user.getRole().toString());
+            }
+        }
+        String newPassword = request.getParameter("password");
```
this code I've already seen
DRY - don't repeat yourself

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/NewPasswordServlet.java:
```diff
@@ -0,0 +1,105 @@
+        String recoveryToken = request.getParameter("recoveryToken");
+        long userId = userPostService.getRecoveryUserId(recoveryToken);
+        Optional<User> optionalUser = userService.findById(userId);
+        if (optionalUser.isPresent()){
```
in servlet should be only one call to service

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/NewPasswordServlet.java:
```diff
@@ -0,0 +1,105 @@
+            User user = optionalUser.get();
+            if (newPassword.isEmpty()) {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
```
why check input data after calling service and dao?

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/web/servlets/RecoveryPasswordServlet.java:
```diff
@@ -0,0 +1,103 @@
+                attributes.put("userRole", user.getRole().toString());
+            }
+        }
+        PageGenerator.getInstance().process("recoveryPassword", attributes, response.getWriter());
```
DRY

Rooman commented on src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDao.java:
```diff
@@ -150,4 +153,24 @@ public boolean update(String email, User user) {
+                }
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
```
@SneakThrows

TarasOrlovskyi replied:
I will add logs and therefore refuse to use this annotation. Thanks for the comment

Rooman replied:
I think logs is unnecessary here, and @SneakyThrows is much better solution

AlexKievDeveloper replied:
> I think logs is unnecessary here, and @SneakyThrows is much better solution


As @SneakyThrows result after compilation we will have try with:
catch (Throwable var11) {
            throw var11;
        }
As I understood if SqlException throwed, we will catch it and re-throw to the ErrorHandler through service and web levels. Throwing sql exception through web level it is normal behaviour?
And now we use try with resources for closing connection and prepared statement, if we add @SneakyThrows should we use them without try with resources?

TarasOrlovskyi committed:
Change after last review and added tests

1. Added entity RecoveryToken
2. Added RecoveryRowMapper
3. Created RecoveryPasswordDao and implementation
   DefaultRecoveryPasswordDao
4. If during adding recovery token in db occurs conflict then this row
   will be update. Changed query to db
5. UserPostDao returned to its previous state
6. Added logs into findById method in JdbcUserDao
7. Deleted unecessary methods from RecoveryPasswordDao - update and
   findByUserId
8. Into RecoveryPasswordService added UserService in order to take it
   away from Servlets
9. From addRecoveryUserToken removed update toker logic and transfered
   it to Dao level (resolve conflict when add token by update)
10. In ChangePasswordServlet swapped conditions and removed UserService.
    Added condition if newPassword is null. Added condition if token lifetime is ended then
    token will be remove and user see special message
11. In RecoveryPasswordServlet removed UserService, added condition if
    email is null.
12. In flyway migration added that token is unique and added
    token_lifetime column
13. Added tests for RecoveryRowMapper, JdbcRecoveryPasswordDao,
    DefaultRecoveryPasswordService, ChangePasswordServlet,
    RecoveryPasswordServlet
